File: core/src/nl/t64/game/rpg/screens/loot/LootSlotTaker.java
package nl.t64.game.rpg.screens.loot;

import nl.t64.game.rpg.Utils;
import nl.t64.game.rpg.audio.AudioCommand;
import nl.t64.game.rpg.audio.AudioEvent;
import nl.t64.game.rpg.components.party.InventoryItem;
import nl.t64.game.rpg.screens.inventory.itemslot.ItemSlot;
import nl.t64.game.rpg.screens.inventory.itemslot.ItemSlotSelector;
import nl.t64.game.rpg.screens.inventory.messagedialog.MessageDialog;


class LootSlotTaker {

    private final ItemSlotSelector selector;

    private ItemSlot sourceSlot;

    LootSlotTaker(ItemSlotSelector selector) {
        this.selector = selector;
    }

    void take(ItemSlot itemSlot) {
        this.sourceSlot = itemSlot;
        tryPutLootSlotToInventorySlot();
    }

    private void tryPutLootSlotToInventorySlot() {
        sourceSlot.getPossibleInventoryImage()
                  .map(inventoryImage -> inventoryImage.inventoryItem)
                  .ifPresent(this::tryPutLootSlotToInventorySlot);
    }

    private void tryPutLootSlotToInventorySlot(InventoryItem candidateItem) {
        if (itemIsResource(candidateItem) || Utils.getGameData().getInventory().hasEmptySlot()) {
            putLootSlotToInventorySlot(candidateItem);
        } else {
            new MessageDialog("Inventory is full.").show(sourceSlot.getStage(), AudioEvent.SE_MENU_ERROR);
        }
    }

    private boolean itemIsResource(InventoryItem candidateItem) {
        return candidateItem.isStackable()
               && Utils.getGameData().getInventory().hasRoomForResource(candidateItem.getId());
    }

    private void putLootSlotToInventorySlot(InventoryItem candidateItem) {
        Utils.getAudioManager().handle(AudioCommand.SE_PLAY_ONCE, AudioEvent.SE_TAKE);
        Utils.getGameData().getInventory().autoSetItem(candidateItem);
        sourceSlot.clearStack();
        selector.findNextSlot();
    }

}

File: core/src/nl/t64/game/rpg/screens/loot/LootSlotTaker.kt
package nl.t64.game.rpg.screens.loot

import nl.t64.game.rpg.Utils.audioManager
import nl.t64.game.rpg.Utils.gameData
import nl.t64.game.rpg.audio.AudioCommand
import nl.t64.game.rpg.audio.AudioEvent
import nl.t64.game.rpg.components.party.inventory.InventoryItem
import nl.t64.game.rpg.screens.inventory.itemslot.ItemSlot
import nl.t64.game.rpg.screens.inventory.itemslot.ItemSlotSelector
import nl.t64.game.rpg.screens.inventory.messagedialog.MessageDialog


internal class LootSlotTaker(
    private val selector: ItemSlotSelector
) {

    private lateinit var sourceSlot: ItemSlot

    fun take(itemSlot: ItemSlot) {
        sourceSlot = itemSlot
        tryPutLootSlotToInventorySlot()
    }

    private fun tryPutLootSlotToInventorySlot() {
        sourceSlot.getPossibleInventoryImage()?.let {
            tryPutLootSlotToInventorySlot(it.inventoryItem)
        }
    }

    private fun tryPutLootSlotToInventorySlot(candidateItem: InventoryItem) {
        if (itemIsResource(candidateItem) || gameData.inventory.hasEmptySlot()) {
            putLootSlotToInventorySlot(candidateItem)
        } else {
            MessageDialog("Inventory is full.").show(sourceSlot.stage, AudioEvent.SE_MENU_ERROR)
        }
    }

    private fun itemIsResource(candidateItem: InventoryItem): Boolean {
        return (candidateItem.isStackable
                && gameData.inventory.hasRoomForResource(candidateItem.id))
    }

    private fun putLootSlotToInventorySlot(candidateItem: InventoryItem) {
        audioManager.handle(AudioCommand.SE_PLAY_ONCE, AudioEvent.SE_TAKE)
        gameData.inventory.autoSetItem(candidateItem)
        sourceSlot.clearStack()
        selector.findNextSlot()
    }

}
